

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>BAM conversion to other formats &mdash; neatseq_flow 1.1.0 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../genindex.html"/>
        <link rel="search" title="Search" href="../search.html"/>
    <link rel="top" title="neatseq_flow 1.1.0 documentation" href="../index.html"/> 

  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> neatseq_flow
          

          
            
            <img src="../_static/NeatSeq_Flow_logo.png" class="logo" />
          
          </a>

          
            
            
              <div class="version">
                1.1.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <p class="caption"><span class="caption-text">Detailed documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../01.concept.html">NeatSeq-Flow concept</a></li>
<li class="toctree-l1"><a class="reference internal" href="../02.build_WF.html">Building a workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="../03.output.html">Output directory structure</a></li>
<li class="toctree-l1"><a class="reference internal" href="../06.addnew_module.html">For the programmer - Adding modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../07.monitor.html">Monitor docs</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../index.html">neatseq_flow</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          

 



<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../index.html">Docs</a> &raquo;</li>
      
    <li>BAM conversion to other formats</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/modules/others.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="bam-conversion-to-other-formats">
<h1>BAM conversion to other formats<a class="headerlink" href="#bam-conversion-to-other-formats" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="modules-included-in-this-section">
<p class="topic-title first">Modules included in this section</p>
<ul class="simple">
<li><a class="reference internal" href="#genomecoveragebed" id="id22"><code class="docutils literal"><span class="pre">genomeCoverageBed</span></code></a></li>
<li><a class="reference internal" href="#ucsc-bw-wig" id="id23"><code class="docutils literal"><span class="pre">UCSC_BW_wig</span></code></a></li>
<li><a class="reference internal" href="#igv-count" id="id24"><code class="docutils literal"><span class="pre">IGV_count</span></code></a></li>
<li><a class="reference internal" href="#igv-totdf" id="id25"><code class="docutils literal"><span class="pre">IGV_toTDF</span></code></a></li>
</ul>
</div>
<span class="target" id="module-step_classes.mapping.genomeCoverageBed"></span><div class="section" id="genomecoveragebed">
<h2><a class="toc-backref" href="#id22"><code class="docutils literal"><span class="pre">genomeCoverageBed</span></code></a><a class="headerlink" href="#genomecoveragebed" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running bedtools genomecov:</p>
<p>The module builds a bedgraph (bdg) file based on an existing BAM file.</p>
<div class="section" id="requires">
<h3>Requires<a class="headerlink" href="#requires" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">BAM file in the following slot:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;bam&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="output">
<h3>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h3>
<ul>
<li><dl class="first docutils">
<dt>Puts output BedGraph files in the following slots:</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;bdg&quot;]</span></code></li>
</ul>
</dd>
</dl>
</li>
</ul>
</div>
<div class="section" id="parameters-that-can-be-set">
<h3>Parameters that can be set<a class="headerlink" href="#parameters-that-can-be-set" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>-g</td>
<td>path to chrom.sizes</td>
<td>You must redirect the -g parameter. Create the chrom.sizes file for the reference genome with <code class="docutils literal"><span class="pre">samtools</span> <span class="pre">faidx</span></code> followed by <code class="docutils literal"><span class="pre">cut</span> <span class="pre">-f1,2</span></code>.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="lines-for-parameter-file">
<h3>Lines for parameter file<a class="headerlink" href="#lines-for-parameter-file" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">genCovBed_bwt1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">genomeCoverageBed</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">sam_bwt1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">bedtools</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">genomeCoverageBed</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">-</span><span class="n">bg</span><span class="p">:</span> 
        <span class="o">-</span><span class="n">g</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">ref_genome</span><span class="o">/</span><span class="n">ref_genome</span><span class="o">.</span><span class="n">chrom</span><span class="o">.</span><span class="n">sizes</span>
</pre></div>
</div>
</div>
<div class="section" id="references">
<h3>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<span class="target" id="module-step_classes.UCSCtools.UCSC_BW_wig"></span><div class="section" id="ucsc-bw-wig">
<h2><a class="toc-backref" href="#id23"><code class="docutils literal"><span class="pre">UCSC_BW_wig</span></code></a><a class="headerlink" href="#ucsc-bw-wig" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for creating wig and bigwig files using UCSC tools:</p>
<p>The module creates bigwig and wig files from the current active BedGraph file.</p>
<div class="section" id="id1">
<h3>Requires<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">BedGraph file in the following slot:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;bdg&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id2">
<h3>Output<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts output sam files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li>self.sample_data[&lt;sample&gt;][“bw”]</li>
<li>self.sample_data[&lt;sample&gt;][“wig”]</li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id3">
<h3>Parameters that can be set<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>bedGraphToBigWig_params</td>
<td><em>e.g.</em> -blockSize=10 -itemsPerSlot=20</td>
<td>Parameters to pass to <code class="docutils literal"><span class="pre">bedGraphToBigWig</span></code></td>
</tr>
<tr class="row-odd"><td>bigWigToWig_params</td>
<td><em>e.g.</em> -chrom X1 -start X2 -end X3</td>
<td>Parameters to pass to <code class="docutils literal"><span class="pre">bigWigToWig</span></code></td>
</tr>
<tr class="row-even"><td>script_path</td>
<td>&#160;</td>
<td>Path to dir where UCSC tools are located.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Set <code class="docutils literal"><span class="pre">script_path</span></code> to the path ro the UCSC tools, not to a specific tool!!!
Both <code class="docutils literal"><span class="pre">bedGraphToBigWig</span></code> and <code class="docutils literal"><span class="pre">bigWigToWig</span></code> will be executed. To set specific params, use <code class="docutils literal"><span class="pre">bedGraphToBigWig_params</span></code> and <code class="docutils literal"><span class="pre">bigWigToWig_params</span></code>, respectively.</p>
</div>
</div>
<div class="section" id="id4">
<h3>Lines for parameter file<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">UCSCmap_bams</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">UCSC_BW_wig</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">genCovBed</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">kentUtils</span><span class="o">/</span><span class="nb">bin</span>
    <span class="n">bedGraphToBigWig_params</span><span class="p">:</span> <span class="o">-</span><span class="n">blockSize</span> <span class="mi">10</span> <span class="o">-</span><span class="n">itemsPerSlot</span> <span class="mi">20</span>
    <span class="n">bigWigToWig_params</span><span class="p">:</span> <span class="o">-</span><span class="n">chrom</span> <span class="n">X1</span> <span class="o">-</span><span class="n">start</span> <span class="n">X2</span> <span class="o">-</span><span class="n">end</span> <span class="n">X3</span>
    <span class="n">genome</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">genome</span><span class="o">.</span><span class="n">chrom</span><span class="o">.</span><span class="n">sizes</span>
</pre></div>
</div>
</div>
<div class="section" id="id5">
<h3>References<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h3>
<p>Kent, W.J., Sugnet, C.W., Furey, T.S., Roskin, K.M., Pringle, T.H., Zahler, A.M. and Haussler, D., 2002. <strong>The human genome browser at UCSC</strong>. <em>Genome research</em>, 12(6), pp.996-1006.</p>
</div>
</div>
<span class="target" id="module-step_classes.IGVtools.IGV_count"></span><div class="section" id="igv-count">
<h2><a class="toc-backref" href="#id24"><code class="docutils literal"><span class="pre">IGV_count</span></code></a><a class="headerlink" href="#igv-count" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running IGVtools count:</p>
<div class="section" id="id6">
<h3>Requires<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Either SAM or BAM files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;bam&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;sam&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id7">
<h3>Output<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts output tdf or wig files in one the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;wig&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;tdf&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id8">
<h3>Parameters that can be set<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>format</td>
<td>wig|tdf</td>
<td>Determines whether to create a ‘wig’ or ‘tdf’ file.</td>
</tr>
<tr class="row-odd"><td>genome</td>
<td>&#160;</td>
<td>Path to chrom.sizes file for reference genome</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id9">
<h3>Lines for parameter file<a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">IGVcount1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">IGV_count</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">samtools1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="n">java</span> <span class="o">-</span><span class="n">Xmx1500m</span> <span class="o">-</span><span class="n">jar</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">igvtools</span><span class="o">.</span><span class="n">jar</span> <span class="n">count</span>
    <span class="nb">format</span><span class="p">:</span> <span class="n">tdf</span>   <span class="c1"># Options: &#39;tdf&#39; or &#39;wig&#39;</span>
    <span class="n">genome</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">genome</span><span class="o">.</span><span class="n">chrom</span><span class="o">.</span><span class="n">sizes</span>
</pre></div>
</div>
</div>
<div class="section" id="id10">
<h3>References<a class="headerlink" href="#id10" title="Permalink to this headline">¶</a></h3>
<p>Thorvaldsdóttir, H., Robinson, J.T. and Mesirov, J.P., 2013. <strong>Integrative Genomics Viewer (IGV): high-performance genomics data visualization and exploration</strong>. <em>Briefings in bioinformatics</em>, 14(2), pp.178-192.</p>
</div>
</div>
<span class="target" id="module-step_classes.IGVtools.IGV_toTDF"></span><div class="section" id="igv-totdf">
<h2><a class="toc-backref" href="#id25"><code class="docutils literal"><span class="pre">IGV_toTDF</span></code></a><a class="headerlink" href="#igv-totdf" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running IGVtools toIGV:</p>
<div class="section" id="id11">
<h3>Requires<a class="headerlink" href="#id11" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">WIG file in the following slot:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;wig&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id12">
<h3>Output<a class="headerlink" href="#id12" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts output tdf file in one the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;tdf&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id13">
<h3>Parameters that can be set<a class="headerlink" href="#id13" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>genome</td>
<td>&#160;</td>
<td>Path to chrom.sizes file for reference genome</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id14">
<h3>Lines for parameter file<a class="headerlink" href="#id14" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">IGV2TDF</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">IGV_toTDF</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">samtools1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">java</span> <span class="o">-</span><span class="n">Xmx1500m</span> <span class="o">-</span><span class="n">jar</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">igvtools</span><span class="o">.</span><span class="n">jar</span> <span class="n">toTDF</span> 
    <span class="n">genome</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">genome</span><span class="o">.</span><span class="n">chrom</span><span class="o">.</span><span class="n">sizes</span>
</pre></div>
</div>
</div>
<div class="section" id="id15">
<h3>References<a class="headerlink" href="#id15" title="Permalink to this headline">¶</a></h3>
<p>Thorvaldsdóttir, H., Robinson, J.T. and Mesirov, J.P., 2013. <strong>Integrative Genomics Viewer (IGV): high-performance genomics data visualization and exploration</strong>. <em>Briefings in bioinformatics</em>, 14(2), pp.178-192.</p>
</div>
</div>
</div>
<div class="section" id="chip-seq-peak-calling">
<h1>ChIP-seq peak calling<a class="headerlink" href="#chip-seq-peak-calling" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="id16">
<p class="topic-title first">Modules included in this section</p>
<ul class="simple">
<li><a class="reference internal" href="#macs2-callpeak" id="id26"><code class="docutils literal"><span class="pre">macs2_callpeak</span></code></a></li>
</ul>
</div>
<span class="target" id="module-step_classes.ChIP_seq.macs2_callpeak"></span><div class="section" id="macs2-callpeak">
<h2><a class="toc-backref" href="#id26"><code class="docutils literal"><span class="pre">macs2_callpeak</span></code></a><a class="headerlink" href="#macs2-callpeak" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running macs2 callpeak:</p>
<div class="section" id="id17">
<h3>Requires<a class="headerlink" href="#id17" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">bam files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;bam&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If using control (input) samples, make sure you include a sample-control table in your sample file.</p>
</li>
</ul>
</div>
<div class="section" id="id18">
<h3>Output<a class="headerlink" href="#id18" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts output macs2 output files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;prefix&quot;])</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;peak_bed&quot;])</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;peak_xls&quot;])</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;summit_bed&quot;])</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If <code class="docutils literal"><span class="pre">--bdg</span></code> (or <code class="docutils literal"><span class="pre">-B</span></code>) was specified, puts output bdg files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;control_lambda&quot;]</span></code> - Control BedGraph</li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;treat_pileup&quot;]</span></code> - Treatment BedGraph</li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;bdg&quot;]</span></code> - Treatment BedGraph</li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;control&gt;][&quot;bdg&quot;]</span></code> - Control BedGraph</li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If <code class="docutils literal"><span class="pre">bedToBigBed_path</span></code> was specified, puts output bigbed files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;bb&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If <code class="docutils literal"><span class="pre">getfasta</span></code> was specified, puts output fasta files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;peak_fasta&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id19">
<h3>Parameters that can be set<a class="headerlink" href="#id19" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>bedToBigBed_path</td>
<td>path to bedToBigBed</td>
<td>Runs bedToBigBed to convert the peak bed files into bigbed for uploading to UCSC.</td>
</tr>
<tr class="row-odd"><td>chrom.sizes</td>
<td>path to chrom.sizes for reference genome</td>
<td>If running bedToBigBed, you must supply the genome chrom.sizes file.</td>
</tr>
<tr class="row-even"><td>getfasta</td>
<td>&#160;</td>
<td>If set, a fasta file containing the peak sequences will be produced.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id20">
<h3>Lines for parameter file<a class="headerlink" href="#id20" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">macs1_CP</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">macs2_callpeak</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">samtools1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">macs2</span> <span class="n">callpeak</span>
    <span class="n">bedToBigBed_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">kentUtils</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">bedToBigBed</span>
    <span class="n">chrom</span><span class="o">.</span><span class="n">sizes</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">genome</span><span class="o">.</span><span class="n">chrom</span><span class="o">.</span><span class="n">sizes</span>
    <span class="n">getfasta</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">bedtools</span> <span class="n">getfasta</span> <span class="o">-</span><span class="n">name</span> <span class="o">-</span><span class="n">s</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">SPMR</span><span class="p">:</span> 
        <span class="o">--</span><span class="n">bdg</span><span class="p">:</span> 
        <span class="o">-</span><span class="n">g</span><span class="p">:</span>     <span class="n">mm</span>
        <span class="o">--</span><span class="n">bw</span><span class="p">:</span>   <span class="mi">400</span>
</pre></div>
</div>
</div>
<div class="section" id="id21">
<h3>References<a class="headerlink" href="#id21" title="Permalink to this headline">¶</a></h3>
<p>Feng, J., Liu, T., Qin, B., Zhang, Y. and Liu, X.S., 2012. <strong>Identifying ChIP-seq enrichment using MACS</strong>. <em>Nature protocols</em>, 7(9), pp.1728-1740.</p>
</div>
</div>
</div>


           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Menachem Sklarz.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'1.1.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>